common/libs/network_protocol/network_protocol.proto

```diff
@@ -1,128 +1,202 @@
 syntax = "proto2";
 
-enum ControllerMsgType {
-  ACK = 0x10;
-  ERR_CHECK_SUM = 0x11;
-  ERR_MODE = 0x12;
-  ERR_CMD = 0x13;
-  STATUS = 0x20;
-  ALARM = 0x30;
-  DATA = 0x40;
+import "nanopb.proto";
+
+// This file defines messages sent between the ventilator's microcontroller and
+// the computer controlling its GUI.
+//
+// When you modify this file you also need to regenerate the C code.  On my
+// Mac:
+//
+//  $ brew install nanopb-generator
+//  $ pip3 install protobuf
+//  $ cd common/libs/network_protocol
+//  $ protoc
+//  --plugin=/usr/local/Cellar/nanopb-generator/0.4.1/bin/protoc-gen-nanopb \
+//      -I/usr/local/Cellar/nanopb-generator/0.4.1/libexec/proto -I. \
+//      --nanopb_out=. network_protocol.proto
+//
+// A note about optional vs. required fields: Required fields are required; if
+// a protobuf is missing a required field, it fails to deserialize.
+//
+// At Google, required fields are generally considered harmful, as they make it
+// difficult to change a protobuf.  If the sender and receiver can have
+// different versions of the protobuf, you can never add or remove a required
+// field, because then the proto may fail to deserialize for want of a new
+// required field.
+//
+// In this project we use required fields for the simple reason that the nanopb
+// api for required fields is more convenient.  In particular, every optional
+// field is represented as two C struct fields, whereas required fields require
+// only one C struct field:
+//
+//   // proto
+//   messsage Foo {
+//     optional int32 opt = 1;
+//     required int32 req = 2;
+//   }
+//
+//   // generated C code
+//   struct Foo {
+//     bool has_opt;
+//     int32_t opt;
+//     int32_t req;
+//   }
+//
+// Not only do the optional fields take up more RAM, they're also harder to
+// work with, because you always have to check has_opt before reading `opt`,
+// and you always have to set has_opt before writing `opt`.
+//
+// Mismatches between proto versions is not an issue for us; we'll simply
+// ensure that both sides always have the same version.  This is somewhat
+// constraining, but better that than reasoning about what happens when X new
+// field is missing on one side or the other.
+
+// Periodically sent from the GUI to the controller.
```

Comment:
Could you add an overview of the principles of the protocol - that it is an exchange of absolute state rather than commands, the meaning of ACKs (e.g. what constitutes an ACK, how do we handle if an ACK doesn't arrive in time, does every message have to be acked or no) etc?